### ISSUE TITLE
replace libhttpd

### ISSUE THREAD

**Issue by mwarning** [nodogsplash/nodogsplash#46]
libhttpd does not seem to be maintained anymore and lacks quality.
Any ideas for replacing it?

**Comment by mwarning** [comment]
libmicrohttpd might be an option: https://www.gnu.org/software/libmicrohttpd/

**Comment by lynxis** [comment]
also good candidates are libevhtp libneon. All of them are already supported by OpenWrt.

**Comment by lynxis** [comment]
https://github.com/lynxis/nodogsplash/tree/microhttpd
working so far:
* mimetypes
* serving files
* templating

missing:
* auth logic

**Comment by mwarning** [comment]
Great work. Let us know when it works  so we can merge it. :-)

**Comment by lynxis** [comment]
I would like to merge this into a seperate branch and tag all old tickets with version "v0.9" and bugs for the "new" version with "v2". I'm not sure what we should do with v0.9. I think it's a good idea to release a version 1.0.

**Comment by mwarning** [comment]
I agree, let's create release based on the current sources and create a branch for the new code for now.

**Comment by nhimcoi30889** [comment]
Hi @lynxis @mwarning ,
at the moment, i can use css style for nodogsplash ( nodogsplash_0.9_beta9.9.9-6_ramips ) ? 
i tried with setup libevhtp libneon and libmicrohttpd but it dont working :(

**Comment by mwarning** [comment]
@nhimcoi30889 you can use CSS just fine. There should be no problem.

**Comment by bluewavenet** [comment]
... as long as any css files are in /etc/nodogsplash/htdocs

**Comment by nhimcoi30889** [comment]
@mwarning : i link to css ( at /www/mycss ) , but it only show html format .
Can you help me about more detail ?
Thanks bro

**Comment by nhimcoi30889** [comment]
I insert into splash.html as : 
<link href="/etc/nodogsplash/htdocs/images/bootstrap.min.css" type="text/css" rel="stylesheet"></link>
but it dont show css

**Comment by bluewavenet** [comment]
That should be:
`/images/bootstrap.min.css`
not `/etc/nodogsplash/htdocs/images/bootstrap.min.css`

**Comment by nhimcoi30889** [comment]
i added
`
    <link href="/images/bootstrap.min.css" type="text/css" rel="stylesheet">
    <link href="/images/font-awesome.min.css" rel="stylesheet">`
But mobile devices still dont show ssh ( it only show same type="text/html"
I can miss lib ? i installed libevhtp libneon and libmicrohttpd

**Comment by nhimcoi30889** [comment]
if i put my css on /www/mycss/bootstrap.min.css
i edited link to bellow : 
`
    <link href="http://ipaddress_openwrt/mycss/bootstrap.min.css" type="text/css" rel="stylesheet">`
And it still dont show :(

**Comment by bluewavenet** [comment]
The web server built into Nodogsplash is very simple and basic. Fine for a simple splash page.
Your problem here is not anything to do with issue #46
Why do you need separate css files for a simple page? If you are trying to do something much more sophisticated you would be better of using the uhttpd already running on OpenWrt. This is by default not accessible to clients when at the preauth stage. You need to allow access by adding an entry into nodogsplash.conf

With your css files under http://ipaddress_openwrt/mycss/bootstrap.min.css, add the following line to nodogsplash.conf:
`     # FirewallRuleSet: preauthenticated-users`
`FirewallRule allow tcp port 80 to [ipaddress_openwrt]`

You do not need libevhtp libneon or libmicrohttpd .

**Comment by nhimcoi30889** [comment]
Tks @bluewavenet ,
I need to use css for some options background
i added `FirewallRule allow tcp port 80 to [ipaddress_openwrt]` to nodogsplash.conf file but dont work
From ver 14.09, it show normally, but after update to 15.05.1 it dont working :( 
Can you give some solution for this problem ? 
I can give all config files. 
i have never use uhttpd for this case

**Comment by bluewavenet** [comment]
For simplicity do the following:
1. 
[nodogsplash.init.d.txt](https://github.com/nodogsplash/nodogsplash/files/275863/nodogsplash.init.d.txt)
Replace /etc/init.d/nodogsplash with this file, renaming to nodogsplash.

2. Put  this version of nodogsplash.conf into /etc/nodogsplash/ and edit your router's ip address accordingly.
[nodogsplash.conf.txt](https://github.com/nodogsplash/nodogsplash/files/275864/nodogsplash.conf.txt)

Then in a ssh terminal type:
/etc/init.d/nodogsplash enable
/etc/init.d/nodogsplash start

This should do the trick, I use this all the time.

You may find nodog fails to start on rebooting, if so I have a fix for that too, so let me know.

**Comment by lynxis** [comment]
@nhimcoi30889  I don't think you used nodogsplash 2.0. nodogsplash 2 doesn't have any problems with that.

**Comment by bluewavenet** [comment]
It would be nice if version 2 was released, or is it just a well kept secret? ;)

**Comment by lynxis** [comment]
@bluewavenet it's here on github. we need to create two releases (v1 old codebase, v2 libmicrohttp). it's here in a seperate branch (v2)

**Comment by nhimcoi30889** [comment]
@lynxis i think not yet v2
I use nodog newest version at this link
'https://downloads.openwrt.org/chaos_calmer/15.05/ramips/mt7620/packages/routing/nodogsplash_0.9_beta9.9.9-5_ramips_24kec.ipk`

**Comment by nhimcoi30889** [comment]
I think nodog for ramips not yet support version 2, because i only find newest is 0.9 beta 9.9.9-5
So, i can use css

**Comment by lynxis** [comment]
neither lede nor openwrt got a v2. I plan to do releases next week.

 * version 1.0 our own httpd implementation
 * version 2.0 libmicrohttpd

but to do the releases I want to do some automatic tests using travis

**Comment by nhimcoi30889** [comment]
@lynxis can you tell me when do you release ?
Thanks bro

**Comment by nhimcoi30889** [comment]
Hi @lynxis ,
i found a nodogsplash as you said at this : 
`https://github.com/nodogsplash/nodogsplash/tree/v2`
i want to install on my router ( running openwrt 15.05 ) so it's working normally ?
example i want to use it as a .ipk package

**Comment by nhimcoi30889** [comment]
i had installed libmicrohttpd. Can you tell me that i need change some files for use css ?
Tks ,

**Comment by lynxis** [comment]
you have to compile it using the LEDE/OpenWrt buildsystem to get ipk.

**Comment by bluewavenet** [comment]
@nhimcoi30889 
Follow the instructions I posted here earlier (replacing nodogsplash init.d script and nodogsplash.conf).
This works perfectly and you do not need to install any other software, working with the current OpenWrt release.

**Comment by nhimcoi30889** [comment]
@bluewavenet . Thank bro,
This is my file, you can review it ?
My ip router is 172.15.1.254
[nodogsplash.conf.txt](https://github.com/nodogsplash/nodogsplash/files/276415/nodogsplash.conf.txt)
[nodogsplash.init.d.txt](https://github.com/nodogsplash/nodogsplash/files/276416/nodogsplash.init.d.txt)

**Comment by nhimcoi30889** [comment]
@bluewavenet OMG, it works perfectly. Thanks bro, What is this problem ?

**Comment by bluewavenet** [comment]
The only issue is that Nodogsplash gets the mime type wrong for a .css file.
This should be fixed in v2 as mentioned at the beginning of this thread. You are now using OpenWrt's uhttpd for your .css files.

**Comment by nhimcoi30889** [comment]
@bluewavenet Thks bro,
So, if use uhttpd that affect to system ? like slow performance, slow responsive ...

**Comment by bluewavenet** [comment]
uhttpd was chosen by the developers of Openwrt as it is a full function web server that is very efficient and compact. Using it is the best thing to do if you want more than the simple splash page, splash.html as nothing needs to be installed to use it.

For example I have systems running with more than 300 client logins per day with the authentication pages handled by php running on uhttpd with tokens passed back to nodogsplash.

Nodogsplash requires its own http service on a different port as a consequence of how it works for redirection and authentication. In the current release of Nodogsplash the http library used is old and outdated (libhttpd). Replacing it in Nodogsplash is what this thread is all about and is happening in v2.

If you are having long delays for clients before getting the splash page, see #113

**Comment by nhimcoi30889** [comment]
Oh tks bro @bluewavenet , i feel its fast
But i have 1 problem with your new file. If i use my old file. When client disconnect ( turn off wifi ) with access point. About 2mins, it'll remove out client list ( ndsctl status ) and when turn on wifi, system require pop-up authent. But i change by your file, system very fast but client disconnect dont remove out client list. And it can pop-up re-authent
Can you help me why ?

**Comment by nhimcoi30889** [comment]
Or can you tell me that i need change a little infor on my file attach so can use uhttpd as your file and dont change other everything :)

**Comment by bluewavenet** [comment]
You just need to set "client idle timeout" to whatever you require. The default is 10 minutes. In your file you have set it to 2 minutes but it is commented out. I usually set this to 60 minutes in live systems otherwise it becomes very annoying to clients. 2 minutes will be far too short. Just think if you are a client on your system and writing an email - You stop to think what to write and before you know it you are logged out again.

**Comment by nhimcoi30889** [comment]
Oh, you'r right. i'll set to abount 30 mins. 
and i found it at 2 file
 - /etc/nodogsplash/nodogsplash.conf --- clientidletimeout 2
 - /etc/config/nodogsplash --- option idletimeout ' 2'
I will change on both or only one ?

**Comment by bluewavenet** [comment]
For simplicity you should use BOTH of the files I uploaded. This would mean /etc/config/nodogsplash is not used.

**Comment by nhimcoi30889** [comment]
Ah,
This is : NDS_CONF=/etc/nodogsplash/nodogsplash.conf ==> it use nodogsplash.conf and not use /etc/config/nodogsplash ???

**Comment by bluewavenet** [comment]
It is determined by the init.d script. My script makes it use /etc/nodogsplash/nodogsplash.conf like it used to in earlier versions. The latest version creates a new config file in the /tmp directory every time it starts up. 
I did this for simplicity when upgrading.

**Comment by nhimcoi30889** [comment]
OK, its very perfect. Thank bro
my issues resolved :)
Can you give me your contact ?

**Comment by mwarning** [comment]
@bluewavenet @lynxis: I will try to do a nds 1.0 release on the weekend. Any ideas what changes would be good to have in there?

**Comment by bluewavenet** [comment]
:)
0.9 beta 9.9.9-5 all seems rock solid in terms of reliability and performance. What changes are proposed? #80 and #81?
Is #112 related to the use of libhttpd?  Am I correct in assuming v1 is staying with libhttpd?
Is seems #112 has come from  Wifidog perhaps. [http://dev.wifidog.org/ticket/660](url)

**Comment by lynxis** [comment]
it's merged now.

**Comment by Onuonga** [comment]
I can not seem to trace the branch with version 2 and from the releases I can only see version 1.

**Comment by mwarning** [comment]
@Onuonga the version 2 does not have gotten a release yet. But you can compile it yourself.

**Comment by alissonrodrigo** [comment]
@bluewavenet 
good night, I'm starting now with openwrt + nodogsplash. Can you help me? It is configured and running perfectly with the default home screen ... I would like to know if it is possible to put a frame pulling image or web images? if yes what procedure you indicate me. thankful

**Comment by bluewavenet** [comment]
@alissonrodrigo 
NDS is configured to have restricted web server functionality to deliver just the simple splash page.
If you want something more sophisticated you will need to use FAS
See:
https://nodogsplash.readthedocs.io/en/latest/authentication.html#nodogsplash-all-versions-forwarding-authentication-service-fas
and
https://github.com/nodogsplash/nodogsplash/tree/master/forward_authentication_service
This contains a fully functional set of php scripts by way of example.
If you follow the documentation exactly and get the example working you can then go from there and write your own scripts.
Be aware though, on user devices, the Captive Portal Detection (CPD), for security reasons, usually invokes a functionally restricted browser that will inhibit downloads and have javascript disabled. In addition some implementations of CPD immediately close this browser as soon as an Internet connection is found to be available after authentication.

In summary, don't use javascript as most likely the user's device will have it disabled and don't expect your page to stay open on the client device after authentication.

**Comment by alissonrodrigo** [comment]
very good @bluewavenet , thank you worked 'as expected' only one thing that is happening all see that I will try to access Luci OpenWrt gets on the splash screen in a loop I can not access Luci

**Comment by bluewavenet** [comment]
@alissonrodrigo 
To get access to Luci, you will also need to allow access, users to router to port 80 (default), or whatever port you have uhttpd configured for.

**Comment by alissonrodrigo** [comment]
tanks...
